packages/lib-classifier/src/plugins/drawingTools/experimental/models/marks/FreehandLine/FreehandLine.js

```diff
@@ -3,278 +3,657 @@ import { FreehandLine as FreehandLineComponent } from '@plugins/drawingTools/com
+      } else if (action.type == 'splice-end') {
+        if (action.throughBeginning) {
+          let spliceIndex = Math.abs(action.pointIndexDrag - action.pointIndexClose)
+
+          for (let i = spliceIndex; i < self.points.length - 1; i++) {
+            self.splicePoints.push(self.points[i])
+          }
+
+          if (action.isReversed) {
+            self.splicePoints = self.splicePoints.reverse()
+          }
+
+          // we add 1 because we don't splice out either drag/close point
+          self.points.splice(spliceIndex + 1, self.points.length - spliceIndex, ...action.pointsFromClose)
+          self.points.splice(0, 0, ...action.pointsFromDrag)
+        } else {
+          let spliceIndex = (action.isReversed)
+            ? action.pointIndexClose
+            : action.pointIndexDrag
+
+          for (let i = spliceIndex; i <= spliceIndex + action.pointsSpliceCount; i++) {
+            self.splicePoints.push(self.points[i])
+          }
+
+          if (action.isReversed) {
+            self.splicePoints = self.splicePoints.reverse()
+          }
+
+          // we add 1 because we don't splice out either drag/close point
+          self.points.splice(spliceIndex + 1, action.pointsSpliceCount, ...action.pointsFromDrag)
+        }
+        self.spliceActive = true
+        self.spliceReverse = action.isReversed
+        self.spliceThroughBeginning = action.throughBeginning
+        self.spliceDragPointIndex = action.pointIndexDrag
+        self.spliceClosePointIndex = action.pointIndexClose
+
+        self.setDragPoint(self.splicePoints.at(-1))
+        self.setClosePoint(self.points[self.spliceClosePointIndex])
+
+        self.drawActions.pop()
       }
+
+      self.redoActions.push(action)
     },
 
-    revertEdits() {
-      self.cancelClipPath()
-      self.dragPoint = null
-      self.targetPoint = null
+    redo() {
+      if (self.redoActions.length == 0) return
+
+      let action = toJS(self.redoActions.at(-1))
+
+      if (action.type == 'end') {
+        self.closePath()
+      } else if (action.type == 'append') {
+        self.points.splice(self.points.length, 0, ...action.pointsFromDrag)
+        self.setDragPoint(self.points.at(-1))
+        self.drawActions.push(action)
+      } else if (action.type == 'splice-drag-point') {
+        self.spliceDragPointIndex = action.pointIndexDrag
+        self.setDragPoint(self.points[action.pointIndexDrag])
+        self.setClosePoint(null)
+        self.drawActions.push(action)
+      } else if (action.type == 'splice-close-point') {
+        self.setClosePoint(self.points[action.pointIndexDrag])
+        self.drawActions.push(action)
+        self.splicePathSetup()
+      } else if (action.type == 'splice-append') {
+        // complement the redo of appendPathStart
+        if (self.splicePoints.length == 0) {
+          self.splicePoints.push(self.dragPoint)
+        }
+
+        self.splicePoints.splice(self.splicePoints.length, 0, ...action.pointsFromDrag)
```

Comment:
I think you can use concatenation here too:
```suggestion
        self.splicePoints = self.splicePoints.concat(action.pointsFromDrag)
```

Reply:
Same as above - `splice()` manipulates the array itself and likely performs `concat()` under the hood.

Reply:
True, but I had to read this line twice before I understood what it's doing. `.concat` might make it easier to follow what's going on.

Spread operators are fast, so the most readable code might be spreading the two arrays into a new array.
```suggestion
        self.splicePoints = [...self.splicePoints, ...action.pointsFromDrag]
```